File: c++/001-TwoSum-2.cpp
#include<iostream>
#include<vector>
using namespace std;

struct TNum {
    int number;
    int index;
};

bool compare(const TNum &x, const TNum &y) {
    return x.number < y.number;
}

class Solution {
public:
    vector<int> twoSum(vector<int> &numbers, int target) {
        vector<int> ans;
        if (numbers.size() < 2) return ans;
        
        int n = numbers.size();
        vector<TNum> nums(n);
        for (int i=0; i<n; i++) {
            nums[i].number = numbers[i];
            nums[i].index = i;
        }
        std::sort(nums.begin(), nums.end(), compare);
        
        int i = 0, j = n - 1;
        while (i < j) {
            long long sum = (long long)nums[i].number + nums[j].number;
            if (sum == target) {
                int index1 = nums[i].index + 1;
                int index2 = nums[j].index + 1;
                if (index1 > index2) {
                    int temp = index1; index1 = index2; index2 = temp;
                }
                ans.push_back(index1);
                ans.push_back(index2);  
                return ans;
            } else 
            if (sum < target) {
                i++;
            } else {
                j--;
            }
        }
        return ans;
    }
};

int main() {
	return 0;
}

File: python/001-TwoSum-2.py
class Solution:
    class Item :
        def __init__(self, index, value) :
            self.index = index
            self.value = value        
        def __lt__(self, other):
            return self.cmp(other) < 0
        def __gt__(self, other):
            return self.cmp(other) > 0
        def __eq__(self, other):
            return self.cmp(other) == 0
        def __le__(self, other):
            return self.cmp(other) <= 0
        def __ge__(self, other):
            return self.cmp(other) >= 0
        def __ne__(self, other):
            return self.cmp(other) != 0				
        def cmp(self, other) :
            if other.value < self.value or \
                other.value == self.value and other.index < self.index :
                return 1
            elif other.value > self.value or \
                other.value == self.value and other.index > self.index :
                return -1
            else :
                return 0        
				
    # @param {integer[]} nums
    # @param {integer} target
    # @return {integer[]}
    def twoSum(self, nums, target):
        pairList = [Solution.Item(index, value) for index, value in enumerate(nums)]
        pairList.sort()
        i = 0
        j = len(pairList) - 1
        while i < j :
            sum = pairList[i].value + pairList[j].value
            if sum < target :
                i = i + 1
            elif sum > target :
                j = j - 1
            else :
                return min(pairList[i].index, pairList[j].index) + 1, \
                    max(pairList[i].index, pairList[j].index) + 1
				
nums = [2, 7, 11, 15]
target = 9
solution = Solution()
print(solution.twoSum(nums, target))
			